File: ThunderBasics/UIWindow+VisibleViewController.m
//
//  UIWindow+VisibleViewController.m
//  ThunderBasics
//
//  Created by Simon Mitchell on 15/10/2015.
//  Copyright © 2015 threesidedcube. All rights reserved.
//

#import "UIWindow+VisibleViewController.h"

typedef void (^TSCNavigationViewControllerRecursion)(UIViewController *visibleViewController, UINavigationController *navigationController, BOOL *stop);

@implementation UIWindow (VisibleViewController)

- (UIViewController *)visibleViewController
{
    UIViewController *rootViewController = self.rootViewController;
    
    if (rootViewController.splitViewController) {
        
        // If any of the split view controller's viewControllers are presnting work up their view hierarcy
        for (UIViewController *splitViewController in rootViewController.splitViewController.viewControllers) {
            
            if (splitViewController.presentedViewController) {
                return [self visibleViewControllerForViewController:splitViewController.presentedViewController];
            }
        }
        
        // Otherwise navigate through the last view controller on the splitViewController
        return [self visibleViewControllerForViewController:rootViewController.splitViewController.viewControllers.lastObject];
    }

    return [self visibleViewControllerForViewController:rootViewController];
}

- (UIViewController *)visibleViewControllerForViewController:(UIViewController *)viewController
{
    // Work up the view hierarchy if presented
    if (viewController.presentedViewController) {
        return [self visibleViewControllerForViewController:viewController.presentedViewController];
    }
    
    if ([viewController isKindOfClass:[UINavigationController class]]) {
        
        UINavigationController *navigationController = (UINavigationController *)viewController;
        UIViewController *lastViewController = navigationController.viewControllers.lastObject;
        
        return [self visibleViewControllerForViewController:lastViewController];
        
    } else if ([viewController isKindOfClass:[UITabBarController class]]) {
        
        UITabBarController *tabBarController = (UITabBarController *)viewController;
        return [self visibleViewControllerForViewController:tabBarController.selectedViewController];
        
    } else if ([viewController isKindOfClass:[UIViewController class]]) {
        
        return viewController;
        
    }
    
    return viewController;
}

@end

File: ThunderBasicsSwift/UIWindow+VisibleViewController.swift
//
//  UIWindow+VisibleViewController.swift
//  ThunderBasics
//
//  Created by Simon Mitchell on 06/11/2018.
//  Copyright © 2018 threesidedcube. All rights reserved.
//

import UIKit

extension UIWindow {
    
    /// Returns the topmost view controller (i.e. the one visible to the user)
    public var visibleViewController: UIViewController? {
        
        guard let rootViewController = rootViewController else {
            return nil
        }
        
        guard let splitViewController = rootViewController.splitViewController else {
            return visibleViewController(in: rootViewController)
        }
        
        // If none of the split view controller's view controllers are presenting
        guard let presentedViewController = splitViewController.viewControllers.first(where: { $0.presentedViewController != nil })?.presentedViewController else {
            guard let lastViewController = splitViewController.viewControllers.last else {
                return nil
            }
            return visibleViewController(in: lastViewController)
        }
        
        // If any of the split view controller's viewControllers are presenting, work up their heirarchy
        return visibleViewController(in: presentedViewController)
    }
    
    private func visibleViewController(in viewController: UIViewController) -> UIViewController? {
        
        // Work up the view hierarchy if presenting
        if let presentedViewController = viewController.presentedViewController {
            return visibleViewController(in: presentedViewController)
        }
        
        switch viewController {
        case let navigationController as UINavigationController:
            guard let lastViewController = navigationController.viewControllers.last else {
                return navigationController
            }
            return visibleViewController(in: lastViewController)
        case let tabBarController as UITabBarController:
            guard let selectedViewController = tabBarController.selectedViewController else {
                return tabBarController
            }
            return visibleViewController(in: selectedViewController)
        default:
            return viewController
        }
    }
}
